File: JavaForPythonLearner/fraction.py

"""
「クラスに定義してあれば上書き、なければ object のデフォルトを使う」 というのがPythonのルールです。
"""
class Fraction:

    def __init__(self, num, den):
        """
        :param num: The top of the fraction 分子
        :param den: The bottom of the fraction 分母
        """
        self.num = num
        self.den = den

    def __repr__(self):
        if self.num > self.den:
            retWhole = int(self.num / self.den)
            retNum = self.num - (retWhole * self.den)
            return str(retWhole) + " " + str(retNum) + "/" + str(self.den)
        else:
            return str(self.num) + "/" + str(self.den) #type: ignore

    def show(self):
        print(self.num, "/", self.den)
    
    def __add__(self, other):
        other = self.toFract(other) # type: ignore 
        newnum = self.num * other.den + self.den * other.num #type: ignore
        newden = self.den * other.den #type: ignore
        common = gcd(newnum, nuwden) # type: ignore
        return Fraction(int(newnum / common), int(newden / commmon)) # type: ignore
    
    __radd__ = __add__

    def __lt__(self, other):
        num1 = self.num * other.den
        num2 = self.den * other.num
        return num1 < num2
    
    def toFract(self, n):
        if isinstance(n, int):
            other = Fraction(n, 1)
        elif isinstance(n, float):
            wholePart = int(n)
            fracPart = n - wholePart
            # convert to 100ths???
            fracNum = int(fracPart * 100)
            newNum = wholePart * 100 + fracNum
            other = Fraction(newNum, 100)
        elif isinstance(n, Fraction):
            other = n
        else:
            print("Error: cannot add a fraction to a ", type(n))
            return None
        return other
            
def gcd(m, n):
    """
    A helper function for Fraction
    """
    while m % n != 0:
        oldm = m
        oldn = n
        m = oldn
        n = oldm % oldn
    return n

print(sorted([Fraction(5, 16), Fraction(3, 16), Fraction(1, 16) + 1]))

File: JavaForPythonLearner/Fraction.java
// Number (abstract：抽象クラス)
// ├── Integer
// ├── Double
// ├── Float
// └── Fraction ← 子供クラスになる
// Fractionクラスを Integerとか Doubleと同じ階層で扱うことができる

// FractionクラスはNumberで使うメソッドを中で使える
// longValue
// intValue
// floatValue
// doubleValue

import java.util.ArrayList;
import java.util.Collections;

public class Fraction extends Number implements Compareable<Fraction> {
    // privateで宣言されているからゲッターメソッドでアクセスする
    
    private Integer numerator;
    private Integer denominator;

    // コンストラクタ定義
    // インスタンス変数定義：python の self = java の this
    // Javaではすべてのコードをクラス内に記述する必要があるため、「メソッド」しかありません
    public Fraction(Integer num, Integer den) {
        this.numerator = num;
        this.denominator = den;
    }
    // Javaでは、メソッドはシグネチャによって識別されます。メソッドのシグネチャには、メソッド名とすべてのパラメータの型が含まれます。名前とパラメータの型は、Javaコンパイラが実行時にどのメソッドを呼び出すかを判断するのに十分な情報です。
    public Fraction(Integer num) {
        this.numerator = num;
        this.denominator =1;
    }
    public Fraction add(Fraction other) {
        Integer newNum = other.getDenominator() * this.numerator + this.denominator * other.getNumerator();
        Integer newDen = this.denominnator * other.getDenominator();
        Integer common = gcd(newNum, newDen);
        return new Fraction(newNum/common, newDen/common);
    }
    public Fraction add(Integer othrer) {
        return add(new Fraction(other));
    }
    // ゲッター
    // Javaでは、インスタンス変数にゲッターメソッドとセッターメソッドを用意するのが一般的なプログラミング手法です
    public Integer getNumerator() {
        return numerator;
    }
    public void setNumerator(Integer numerator) {
        this.numerator = numerator;
    }
    public Integer getDenominator() {
        return denominator;
    }
    public void setDenominator(Integer denominator) {
        this.denominator = denominator;
    }

    
    public String toString() {
        return numerator.toString() + "/" + denominator.toString();
    }

    public boolean equals(Fraction other) {
        Integer num1 = this.numerator * other.getDenominator();
        Integer num2 = this.denominator * other.getDenominator();
        if (num1 == num2)
            return true;
        else
            return false;
    }
    
    
    private static Integer gcd(Integer m, Integer n) {
        while (m % n !=0) {
            Integer oldm = m;
            Integer oldn = n;
            m = oldn;
            n = oldm%oldn;
        }
        return n;
    }
    public static void main(String[] args) {
        Fraction f1 = new Fraction(1,2);

        System.out.println(f1.add(1));
    }

     public double doubleValue() {
        return numerator.doubleValue() / denominator.doubleValue();
    }

    public float floatValue() {
        return numerator.floatValue() / denominator.floatValue();
    }

    public int intValue() {
        return numerator.intValue() / denominator.intValue();
    }

    public long longValue() {
        return numerator.longValue() / denominator.longValue();
    }

    public int compareTo(Fraction othre) {
        Integer num1 = this.numerator * other.getDenominator();
        Integer num2 = this.denominator * other.getDenominator();
        return num1 - num2;
    }

    private static Integer gcd(Integer m, Integer n) {
        while (m % n != 0) {
            Integer oldm = m;
            Integer oldn = n;
            m = oldn;
            n = oldm%oldn;
        }
        return n;
    }
    public static void main(String[] args) {
        Fraction f1 = new Fraction(1,2);
        Fraction f2 = new = Fraction(2,3);
        Fraction f3 = new Fraction(1,4);

        System.out.println("Adding: " + f1.add(1));
        System.out.println("Calling intValue(): " + f1.intValue(1));
        System.out.println("Calling doubleValue(): " + f1.doubleValue());

        ArrayList<Fraction> myFracs = new ArrayList<Fraction>();
        myFracs.add(f1);
        myFracs.add(f2);
        myFracs.add(f3);
        Collections.sort(myFracs);

        System.out.println("Sorted fractions:");
        for(Fraction f : myFracs) {
            System.out.println(f);
        }
    }
}






